add causal order generator

hroc135 opened the pull request:
とりあえず一段階目としてco = (po + wr)+の定義に従ってcoを生成できるようにしてみました

hroc135 committed:
add causal order generator

shumbo commented on src/c3py/history.py:
```diff
@@ -61,6 +61,20 @@ def __init__(self, data: dict[str, list[Operation]]):
                 self.poset.order_try(f"{process}.{i + 1}", f"{process}.{i + 2}")
+    
+    # co = (po U wr)^+
+    def causal_order(self):
```
```suggestion
    def causal_order(self) -> Self:
```

shumbo replied:
Perhaps you want a separate method for creating `wr` relation?

shumbo replied:
I don't want to add this `causal_order` method on `History` because this method only makes sense for histories of read/write memory. I think the easiest approach would be to make a subclass `WRMemoryHistory` that extends `History` and move this method there.

shumbo commented on src/c3py/history.py:
```diff
@@ -61,6 +61,20 @@ def __init__(self, data: dict[str, list[Operation]]):
+
+            arg = op1.arg
+            ret = op1.ret
+            for id2, op2 in self.label.items():
```
this inner loop makes the function $O(n^2)$ but it should be able to do this in $O(n)$

shumbo commented on src/c3py/history.py:
```diff
@@ -61,6 +61,20 @@ def __init__(self, data: dict[str, list[Operation]]):
+            ret = op1.ret
+            for id2, op2 in self.label.items():
+                if op2.method == "wr" and op2.arg == (arg, ret):
+                    ch.poset.link(id2, id1)
```
It's my fault for not explicitly stating this, but currently, the graph (`G`) in the `Poset` class maintains two invariants (that comes from the definition of strong partial order). 

First, the graph must be acyclic. This is to keep the relation irreflexive.

Second, the graph must be transitive closure. That is, if there is a path from a to b and b to c, there must also be an edge from a to c. 

I'm fine with violating the property temporarily, but I have a feeling that we should NOT return an instance of `Poset` that is actually not a partial order.

I would probably change the function to `def make_co(self) -> Self | None` and return `None` if there is a cycle in the resulting poset. You probably don't need to check every time you add an edge and just need to check once at the end.

You also want to make sure the returned graph is transitive. You should be able to use a library function from `networkx`.

shumbo commented:
I created a new branch, `bad-pattern,`, and changed the base of the PR to that branch because I didn't want to merge the unfinished features to `main`.

hroc135 committed:
create WRMemoryHistory

hroc135 committed:
check if co is acyclic

hroc135 committed:
add test for WRMemoryHistory

hroc135 committed:
moved bad pattern things to bad_pattern.py and bad_pattern_test.py

add is_write_co_init_read method to WRMemoryHistory class

hroc135 committed:
add ThinAirRead checker

hroc135 committed:
add WriteCORead checker

hroc135 committed:
more tests and modify ThinAirRead checker

- add more tests for WRMemoryHistory class
- check ThinAirRead in WRMemoryHistory.make_co

hroc135 committed:
create find_cc_bad_pattern

hroc135 committed:
create find_ccv_bad_pattern

hroc135 committed:
implement cm checker

hroc135 committed:
minor change

remove unnecessary if statement
